apollo-router/src/plugins/cache/entity.rs

```diff
@@ -0,0 +1,617 @@
+    cache_control: CacheControl,
+    cache_key: String,
+) -> Result<(), BoxError> {
+    if let Some(data) = response.response.body().data.as_ref() {
+        let ttl: Option<Duration> = cache_control
+            .ttl()
+            .map(|secs| Duration::from_secs(secs as u64));
+
+        if cache_control.should_store() {
+            cache
+                .insert(
+                    RedisKey(cache_key),
+                    RedisValue(CacheEntry {
+                        control: cache_control,
+                        data: data.clone(),
+                    }),
+                    ttl,
+                )
+                .await;
+        }
+    }
+
+    Ok(())
+}
+
+async fn cache_store_entities_from_response(
+    cache: RedisCacheStorage,
+    response: &mut subgraph::Response,
+    cache_control: CacheControl,
+    mut result_from_cache: Vec<IntermediateResult>,
+) -> Result<(), BoxError> {
+    update_cache_control(&response.context, &cache_control);
+
+    let mut data = response.response.body_mut().data.take();
+
+    if let Some(mut entities) = data
+        .as_mut()
+        .and_then(|v| v.as_object_mut())
+        .and_then(|o| o.remove(ENTITIES))
+    {
+        let new_entities = insert_entities_in_result(
+            entities
+                .as_array_mut()
+                .ok_or_else(|| FetchError::MalformedResponse {
+                    reason: "expected an array of entities".to_string(),
+                })?,
+            &cache,
+            cache_control,
+            &mut result_from_cache,
+        )
+        .await?;
+
+        data.as_mut()
+            .and_then(|v| v.as_object_mut())
+            .map(|o| o.insert(ENTITIES, new_entities.into()));
+        response.response.body_mut().data = data;
+    }
+
+    Ok(())
+}
+
+pub(crate) fn hash_vary_headers(headers: &http::HeaderMap) -> String {
+    let mut digest = Sha256::new();
+
+    for vary_header_value in headers.get_all(header::VARY).into_iter() {
+        if vary_header_value == "*" {
+            return String::from("*");
+        } else {
+            let header_names = match vary_header_value.to_str() {
+                Ok(header_val) => header_val.split(", "),
+                Err(_) => continue,
+            };
+            header_names.for_each(|header_name| {
+                if let Some(header_value) = headers.get(header_name).and_then(|h| h.to_str().ok()) {
+                    digest.update(header_value);
+                    digest.update(&[0u8; 1][..]);
+                }
+            });
+        }
+    }
+
+    hex::encode(digest.finalize().as_slice())
+}
+
+pub(crate) fn hash_query(body: &graphql::Request) -> String {
+    let mut digest = Sha256::new();
+    digest.update(body.query.as_deref().unwrap_or("-").as_bytes());
+    digest.update(&[0u8; 1][..]);
+    digest.update(body.operation_name.as_deref().unwrap_or("-").as_bytes());
+    digest.update(&[0u8; 1][..]);
+
+    hex::encode(digest.finalize().as_slice())
+}
```

Comment:
Suggest using the query hash that is used for studio.
This will enable users to at least discover the query hash if they are connected to graphql.

Reply:
this is a hash of the subgraph query, not the client query. And the hash calculation changes in https://github.com/apollographql/router/pull/4175